### PR TITLE
[SPARK-6474][EC2] Use EC2Connection.run_instances to launch clusters

### PR DESCRIPTION
**Pull request by mrdrozdov**
Instead of using the following call in spark_ec2.py to launch clusters:
- [`ec2.image.Image.run`](http://boto.readthedocs.org/en/latest/ref/ec2.html#boto.ec2.image.Image.run)

this patch uses the equivalent and more current call:
- [`ec2.connection.EC2Connection.run_instances`](http://boto.readthedocs.org/en/latest/ref/ec2.html#boto.ec2.connection.EC2Connection.run_instances)

This is a useful change in that connection.run_instances exposes
additional features such as the capability of handling ebs_optimized
instances, which image.run does not handle.

This improvement was first suggested by @nchammas in a boto issue:
https://github.com/boto/boto/issues/3054

### CODE REVIEW

**Comment by nchammas** [comment]
LGTM.

@mrdrozdov Can you post a few invocations that you used to test things out? Change looks totally good and harmless, but it doesn't hurt to double check (especially since spark-ec2 has no tests apart from `pep8`). :)

**Comment by nchammas** [comment]
Jenkins, test this please. (Though I doubt Jenkins will listen to me.)

**Comment by nchammas** [comment]
Hohoho I guess things have changed around here. :sunglasses:

**Comment by mrdrozdov** [comment]
Congrats on your newfound power @nchammas :grinning:

**Comment by mrdrozdov** [comment]
Launching/Accessing/Destroy clusters using spark-ec2 should be sufficient for this change. Could run the following basic flow to double check that everything's in order:

```
# Basic
./spark-ec2 -k <key_name> -i <key_path> -s 1 launch sparky
./spark-ec2 -k <key_name> -i <key_path> login sparky
./spark-ec2 -k <key_name> -i <key_path> destroy sparky

# Bad AMI (HVM)
./spark-ec2 -k <key_name> -i <key_path> -s 1 -a ami-35b1885c launch sparky

# Custom AMI
./spark-ec2 -k <key_name> -i <key_path> -s 1 --region=us-east-1 --zone=us-east-1d -a ami-5bb18832 launch sparky
```

**Comment by nchammas** [comment]
OK, tried those commands and everything checked out OK. The error when the AMI was bad was a bit verbose, but that's what we had before I guess.

cc @JoshRosen @shivaram

**Comment by shivaram** [comment]
Change looks fine to me but I haven't had a chance to run it yet.

**Comment by mrdrozdov** [comment]
Great. I made a slight change to make reduce the verbosity when given a non-existent AMI. connection.get_image is definitely right call to use here, but there's a slight bug/inconsistency in boto. get_image is supposed to just return null if the AMI doesn't exist, but it ends up throwing an EC2ResponseError which I think was causing the verbosity that @nchammas noticed. I have a PR out for the boto fix, and in the mean time wrapped get_image in a try/except block.

I could have just left the code as before, but I think it's nice to remove the image variable seeing that it's no longer being used, plus get_image is a bit more explicit than get_all_images. Although, it's n2bd.

**Comment by mrdrozdov** [comment]
Jenkins, test this please. (overly optimistic that this might work :stuck_out_tongue_winking_eye:)

**Comment by shivaram** [comment]
Jenkins, test this please

File: ec2/spark_ec2.py [hunk 1]
```diff
@@ -502,8 +502,10 @@ def launch_cluster(conn, opts, cluster_name):
                                 if opts.additional_security_group in (sg.name, sg.id)]
     print "Launching instances..."
 
+    # Check if the AMI exists
     try:
-        image = conn.get_all_images(image_ids=[opts.ami])[0]
+        if conn.get_image(image_id=opts.ami) is None:
+            raise Exception()
```
Comment (nchammas) [inline ec2/spark_ec2.py]: Do we want to include some reason here, or will this make it clear to the user what the error was? (Haven't tested this latest update out yet.)

Comment (nchammas) [inline ec2/spark_ec2.py]: Does this bubble up a clear message to the user about what the problem is?

**Comment by srowen** [comment]
@mrdrozdov is this still relevant and can you address Nicholas's questions?

**Comment by mrdrozdov** [comment]
@srowen I think it is okay to close this. It seems like somewhat of an early optimization and still relies on an unsolved issue in https://github.com/boto/boto.

File: ec2/spark_ec2.py (hunk 1 above)
Comment (mrdrozdov) [inline ec2/spark_ec2.py]: I would say not. The exception handling is hard to follow in this flow due to the bug in boto.
